docker/manager.py

```diff
@@ -0,0 +1,821 @@
+        # Args will have format ARG=foobar
+        if isinstance(updater, str):
+            _args, _, _value = updater.partition("=")
+            args[_args] = _value
+            return
+        elif isinstance(updater, list):
+            for v in updater:
+                update_args_dict(v)
+        elif isinstance(updater, dict):
+            for _args, _value in updater.items():
+                args[_args] = _value
+        else:
+            raise AttributeError(
+                f"cannot add to args dict with unknown type {type(updater)}"
+            )
+        return args
+
+    return update_args_dict
+
+
+def _build_release_tags(build_ctx, release_type):
+    """
+    Our tags will have format: {bentoml_release_version}-{python_version}-{distros}-{image_type}
+    eg:
+    - python3.8-ubuntu20.04-base -> contains conda and python preinstalled
+    - devel-python3.8-ubuntu20.04
+    - 0.13.0-python3.7-centos8-{runtime,cudnn}
+    """
+
+    _python_version = get_data(build_ctx, 'envars', 'PYTHON_VERSION')
+
+    # check if PYTHON_VERSION is set.
+    if not _python_version:
+        raise RuntimeError("PYTHON_VERSION should be set before generating Dockerfile.")
+    if _python_version not in SUPPORTED_PYTHON_VERSION:
+        raise ValueError(
+            f"{_python_version} is not supported. Supported versions: {SUPPORTED_PYTHON_VERSION}"
+        )
+
+    _tag = ""
+    core_string = '-'.join([f"python{_python_version}", build_ctx['add_to_tags']])
+
+    base_tag = '-'.join([core_string, 'base'])
+
+    if release_type == "devel":
+        _tag = "-".join([release_type, core_string])
+    elif release_type == "base":
+        pass
+    else:
+        _tag = "-".join(
+            [
+                get_data(build_ctx, 'envars', 'BENTOML_VERSION'),
+                core_string,
+                release_type,
+            ]
+        )
+    return _tag, base_tag
+
+
+def flatten(arr):
+    for it in arr:
+        if isinstance(it, Iterable) and not isinstance(it, str):
+            for x in flatten(it):
+                yield x
+        else:
+            yield it
+
+
+def load_manifest_yaml(file):
+    with open(file, 'r') as input_file:
+        manifest = yaml.safe_load(input_file)
+
+    v_schema = yaml.safe_load(SPEC_SCHEMA)
+    v = MetadataSpecValidator(
+        v_schema, packages=manifest['packages'], releases=manifest['releases']
+    )
+
+    if not v.validate(manifest):
+        logging.error(f"{file} is invalid. Errors as follow:\n")
+        logging.error(yaml.dump(v.errors, indent=2))
+        exit(1)
+
+    release_spec = v.normalized(manifest)
+    return release_spec
+
+
+def _update_build_ctx(ctx_dict, spec_dict: Dict, *ignore_keys):
+    # flatten release_spec to template context.
+    for k, v in spec_dict.items():
+        if k in list(flatten([*ignore_keys])):
+            continue
+        ctx_dict[k] = v
+
+
+def _add_base_release_type(build_ctx, release_type: list):
```

Comment:
quick hack to setup `base` image, which have conda and python that can be used later to build runtime, devel, etc. Definitely improvements here since I wrote this at 3am 😄